### PR TITLE
add dynamic check of origin for CORS

### PR DESCRIPTION
**Pull request by bentsku**
This is a follow up from #7554, from an idea proposed by @whummer ([in this comment](https://github.com/localstack/localstack/pull/7554#discussion_r1086872025)) , with dynamic handling of CORS origin for web resource served by LocalStack.
This is not mergeable as is, as it still has some logging regarding performance, but I'd like to have some comments / opinions.

### Scenario:

LocalStack is serving a resource meant to be accessed from the browser (CloudFront or S3 website).

But this resource needs to access LocalStack resources as well. Right now, the only way is to set the environment variables of LocalStack to allow the Origin (let’s say, `http://<bucket>.s3-website.localhost.localstack.cloud:4566`).

### Fix

This PR adds a dynamic check of the origin, to validate if it follows some regex matching 2 service domains known for serving content accessible from a user browser (thus enforcing CORS).  We can then add a regex for every resource subdomain in that situation. 

The performance right now is around 4µs on my local machine when a request would be either rejected or matching `s3-website` or `cloudfront`. The performance would not be affected for most requests, as we either don't match CORS because of no `Origin` header, or the request is normally allowed. 

\cc @whummer @dfangl @alexrashed

### CODE REVIEW

**Review by whummer — APPROVED** [review]
Cool! Nice approach that can further streamline the CORS UX. 🎉  

Was able to replicate with the following steps:

1. create a CloudFront distribution, as per our [docs](https://docs.localstack.cloud/user-guide/aws/cloudfront/):
```
$ awslocal s3 mb s3://test
$ echo 'Hello World' > /tmp/hello.txt
$ awslocal s3 cp /tmp/hello.txt s3://test/hello.txt --acl public-read

$ domain=$(awslocal cloudfront create-distribution --origin-domain-name test.s3.amazonaws.com | jq -r '.Distribution.DomainName')
$ echo $domain
ce6554a5.cloudfront.localhost.localstack.cloud
```

2. create a simple `index.html` file that connects to the enpoint:
```
<html>
<script>
fetch('http://ce6554a5.cloudfront.localhost.localstack.cloud:4566/hello.txt');
</script>
</html>
```

3. upload the file to a bucket and expose it as an S3 website:
```
awslocal s3 cp /tmp/index.html s3://test/index.html --content-type text/html
awslocal s3api put-bucket-website --bucket test --website-configuration 'ErrorDocument={Key=index.html},IndexDocument={Suffix=index.html}'
```

4. open http://test.s3-website.localhost.localstack.cloud:4566/index.html in the browser, observe the Network console (request to `ce6554a5.cloudfront.localhost.localstack.cloud:4566`)

Result:
* returns a `403` error on current master  🟥 
* returns `200` and expected file content with the changes in this PR  ✅ 

---
To me, this achieves a decent balance between improving convenience (works out of the box without extra configs), while not sacrificing security (requests still only allowed from `*.localhost.localstack.cloud` pages in the browser). 👍 As usual with CORS or security-critical configurations, would be good to get a second pair of eyes on it, though.. 👀 /cc @dfangl @alexrashed @thrau

File: localstack/aws/handlers/cors.py [hunk 1]
```diff
@@ -75,6 +76,28 @@
 ]
 
 
+def _get_allowed_cors_internal_domains() -> Set[int]:
```
[the remark this answers is not in the document]
Comment (bentsku) [reply localstack/aws/handlers/cors.py:79]: oops, copy pasted 😄

File: localstack/aws/handlers/cors.py [hunk 2]
```diff
@@ -88,9 +111,8 @@ def _get_allowed_cors_origins() -> List[str]:
     ]
     # Add allowed origins for localhost domains, using different protocol/port combinations.
     # If a different port is configured for EDGE_PORT_HTTP, add it to allowed origins as well
-    _ports = set([config.EDGE_PORT] + ([config.EDGE_PORT_HTTP] if config.EDGE_PORT_HTTP else []))
     for protocol in {"http", "https"}:
-        for port in _ports:
+        for port in _get_allowed_cors_ports():
```
[the remark this answers is not in the document]
Comment (bentsku) [reply localstack/aws/handlers/cors.py:115]: That's what I thought and did at first, but then it would be generated only once, and in your other tests where you monkey patch `ALLOWED_CORS_ORIGINS`, you'd have to monkey patch `_ALLOWED_INTERNAL_PORTS` before. I thought it'd make your tests a bit more complicated.

https://github.com/localstack/localstack/blob/be7178da606ec4b15b98bd287fc339ad0c870d0d/tests/unit/test_cors.py#L18-L21

**Comment by bentsku** [comment]
Thanks @whummer for creating a reproducible sample!

Just a note:
> (requests still only allowed from *.localhost.localstack.cloud pages in the browser).
The requests are only allowed from `*.s3-website.localhost.localstack.cloud` and `*.cloudfront.localhost.localstack.cloud` for now. 
Also, the webpage served by those 2 "dynamic origins" would be able to not only access cloudfront, but also for example the SQS API, or any API using the default LocalStack configuration.
Just giving precision for the second pair of eyes 👀

File: localstack/aws/handlers/cors.py [hunk 3]
```diff
@@ -166,11 +196,25 @@ def is_cors_origin_allowed(headers: Headers) -> bool:
         return True
 
     @staticmethod
+    @log_duration(min_ms=0)
```
Comment (thrau) [inline localstack/aws/handlers/cors.py:199]: nit: could we remove this again?

File: localstack/aws/handlers/cors.py [hunk 4]
```diff
@@ -20,6 +20,7 @@
 from localstack.http import Response
 
 from ...constants import LOCALHOST, LOCALHOST_HOSTNAME, PATH_USER_REQUEST
+from ...utils.bootstrap import log_duration
```
Comment (thrau) [inline localstack/aws/handlers/cors.py:23]: nit: ideally we don't use relative imports of modules that are outside the scope of the component we are building. basically: relative imports are used when things belong together in a component, absolute imports are used when importing code that represents a dependency to another component

**Review by thrau — APPROVED** [review]
I think this is fantastic :tada: I don't see a big issue right now :+1:

File: localstack/aws/handlers/cors.py (hunk 4 above)
Comment (bentsku) [reply localstack/aws/handlers/cors.py:23]: Right, I'm not sure why there are dynamic import here, I can change them 👍
